src/freenet/client/async/KeyListenerTracker.java

```diff
@@ -95,61 +102,197 @@ protected static int fixRetryCount(int retryCount) {
+			Logger.minor(this, "Added pending keys to "+this+" : size now "+this.keyListeners.size()+"/"+singleKeyListeners.size()+" : "+listener);
 	}
 	
 	public boolean removePendingKeys(KeyListener listener) {
-		boolean ret;
+		boolean ret = false;
+		byte[] wantedKey = listener.getWantedKey();
+		ByteArrayWrapper wrapper = wantedKey != null ? new ByteArrayWrapper(saltKey(wantedKey)) : null;
 		synchronized (this) {
-			ret = keyListeners.remove(listener);
-			while(logMINOR && keyListeners.remove(listener))
-				Logger.error(this, "Still in pending keys after removal, must be in twice or more: "+listener, new Exception("error"));
+			if(wantedKey != null) {
+				Object o = singleKeyListeners.get(wrapper);
+				if(o == null) {
+					// do nothing
+				} else if(o instanceof KeyListener) {
+					if((ret = (listener == (KeyListener)o)))
+						singleKeyListeners.remove(wrapper);
+				} else {
+					@SuppressWarnings("unchecked")
+					KeyListener[] listeners = (KeyListener[])o;
+					KeyListener[] newListeners = new KeyListener[listeners.length-1];
+					int x = 0;
+					for (KeyListener l : listeners) {
+						if(listener == l) {
+							assert(!ret);
+							ret = true;
+							continue;
+						}
+						if(x == newListeners.length) {
+							assert(!ret);
+							break;
+						}
+						newListeners[x++] = l;
+					}
+					if(!ret) {
+						// do nothing
+					} else {
+						assert(x == newListeners.length);
+						if(newListeners.length == 0) {
+							singleKeyListeners.remove(wrapper);
+						} else if(newListeners.length == 1) {
+							singleKeyListeners.put(wrapper,newListeners[0]);
+						} else {
+							singleKeyListeners.put(wrapper,newListeners);
+						}
+					}
+				}
+			} else {
+				ret = keyListeners.remove(listener);
+			}
 			listener.onRemove();
 		}
 		if (logMINOR)
-			Logger.minor(this, "Removed pending keys from "+this+" : size now "+keyListeners.size()+" : "+listener, new Exception("debug"));
+			Logger.minor(this, "Removed pending keys from "+this+" : size now "+this.keyListeners.size()+"/"+singleKeyListeners.size()+" : "+listener, new Exception("debug"));
 		return ret;
 	}
 	
-	public synchronized boolean removePendingKeys(HasKeyListener hasListener) {
-		boolean found = false;
-		for(Iterator<KeyListener> i = keyListeners.iterator();i.hasNext();) {
-			KeyListener listener = i.next();
-			if(listener == null) {
-				i.remove();
-				Logger.error(this, "Null KeyListener in removePendingKeys()");
-				continue;
+	public boolean removePendingKeys(HasKeyListener hasListener) {
+		boolean ret = false;
+		byte[] wantedKey = hasListener.getWantedKey();
+		ByteArrayWrapper wrapper = wantedKey != null ? new ByteArrayWrapper(saltKey(wantedKey)) : null;
+		synchronized(this) {
+			if(wantedKey != null) {
+				Object o = singleKeyListeners.get(wrapper);
+				if(o == null) {
+					// do nothing
+				} else if(o instanceof KeyListener) {
+					if((ret = ((KeyListener)o).getHasKeyListener() == hasListener)) {
```

Comment:
From Eleriseth: Previous code used "==" too => changing that don't belong to my series,
won't change.